batch/batch/driver/gce.py

```diff
@@ -0,0 +1,147 @@
+
+    async def handle_preempt_event(self, instance, timestamp):
+        await instance.inst_coll.call_delete_instance(instance, 'preempted', timestamp=timestamp)
+
+    async def handle_delete_done_event(self, instance, timestamp):
+        await instance.inst_coll.remove_instance(instance, 'deleted', timestamp)
+
+    async def handle_call_delete_event(self, instance, timestamp):
+        await instance.mark_deleted('deleted', timestamp)
+
+    async def handle_event(self, event):
+        payload = event.get('protoPayload')
+        if payload is None:
+            log.warning('event has no payload')
+            return
+
+        timestamp = dateutil.parser.isoparse(event['timestamp']).timestamp() * 1000
+
+        resource_type = event['resource']['type']
+        if resource_type != 'gce_instance':
+            log.warning(f'unknown event resource type {resource_type}')
+            return
+
+        operation = event.get('operation')
+        if operation is None:
+            # occurs when deleting a worker that does not exist
+            log.info(f'received an event with no operation {json.dumps(event)}')
+            return
+
+        operation_started = operation.get('first', False)
+        if operation_started:
+            event_type = 'STARTED'
+        else:
+            event_type = 'COMPLETED'
+
+        event_subtype = payload['methodName']
+        resource = event['resource']
+        name = parse_resource_name(payload['resourceName'])['name']
+
+        log.info(f'event {resource_type} {event_type} {event_subtype} {name}')
+
+        if not name.startswith(self.machine_name_prefix):
+            log.warning(f'event for unknown machine {name}')
+            return
+
+        if event_subtype == 'v1.compute.instances.insert':
+            if event_type == 'COMPLETED':
+                severity = event['severity']
+                operation_id = event['operation']['id']
+                success = (severity != 'ERROR')
+                self.zone_monitor.zone_success_rate.push(resource['labels']['zone'], operation_id, success)
+        else:
+            instance = self.inst_coll_manager.get_instance(name)
+            if not instance:
+                log.warning(f'event for unknown instance {name}')
+                return
+
+            if event_subtype == 'v1.compute.instances.preempted':
+                log.info(f'event handler: handle preempt {instance}')
+                await self.handle_preempt_event(instance, timestamp)
+            elif event_subtype == 'v1.compute.instances.delete':
+                if event_type == 'COMPLETED':
+                    log.info(f'event handler: delete {instance} done')
+                    await self.handle_delete_done_event(instance, timestamp)
+                elif event_type == 'STARTED':
+                    log.info(f'event handler: handle call delete {instance}')
+                    await self.handle_call_delete_event(instance, timestamp)
+
+    async def event_loop(self):
```

Comment:
I notice that we have this pattern of
```python
async def VERB_CLAUSE_loop(...):
    while True:
        ....
        await asyncio.sleep(TIME)
```
For example: scheduling cancel bump loop, monitor billing limits loop, the worker file cache clean up, the worker disk space monitor, and update region quotas.

I feel like this would be more clear with a new utility function. Perhaps `periodically_call(f, *, period, name)`. I think it's implementation is:
```python
async def periodically_call(period, f, *args, **kwargs):
    async def loop():
        while True:
            await f(*args, **kwargs)
            await asyncio.sleep(period)
    await retry_long_running(f.__name__, loop)
```

This separates the periodicity from the code. We'll see things like:
```
    self.background_tasks.ensure_future(periodically_call(10, self.process_event))
```

A little background: I'm searching for some unifying themes here. I don't like that we have to go hunting across the entire batch codebase to see what periodic loops we have an how often they run. I don't know the full answer yet, but I think this moves in the right direction.

Reply:
It also reduces the volume of indentation in the meaningful code, which I like.